runtime/src/main/java/org/corfudb/runtime/view/stream/StreamAddressSpace.java

```diff
@@ -84,34 +79,30 @@ public boolean contains(long address) {
      * @return returns a as the merged StreamAddressSpace
      */
     public static StreamAddressSpace merge(StreamAddressSpace a, StreamAddressSpace b) {
-        a.mergeFrom(b);
-        return a;
-    }
-
-    // Merge another StreamAddressSpace into this instance
-    private void mergeFrom(StreamAddressSpace other) {
-        this.trimMark = Long.max(trimMark, other.getTrimMark());
-        this.bitmap.or(other.bitmap);
+        a.trimMark = Long.max(a.trimMark, b.getTrimMark());
+        a.bitmap.or(b.bitmap);
 
         // Because the trim mark can increase after the merge another
         // trim needs to be issued
-        this.trim(this.trimMark);
+        a.trim(a.trimMark);
+        return a;
     }
 
     /**
-     * Copy this stream's addresses to a set, under a given boundary (inclusive).
+     * Call a long consumer on range [trimMark + 1, max].
      *
-     * @param maxGlobal maximum address (inclusive upper bound)
+     * @param max maximum address (inclusive upper bound)
      */
-    public NavigableSet<Long> copyAddressesToSet(final long maxGlobal) {
-        NavigableSet<Long> queue = new TreeSet<>();
-        this.bitmap.forEach(address -> {
-            if (address <= maxGlobal) {
-                queue.add(address);
+    public void forEachUpTo(final long max, LongConsumer consumer) {
+        LongIterator iterator = this.bitmap.getLongIterator();
+        while (iterator.hasNext()) {
+            long current = iterator.next();
+            if (current <= max) {
```

Comment:
setTrim can break this contract.

Reply:
are we going to handle this?

Reply:
Yes. Added `(current > trimMark && current <= max)`